lib/expression/callbacks.ex

```diff
@@ -21,81 +21,83 @@ defmodule Expression.Callbacks do
   underscore.
   """
 
-  defmacro __using__(_opts) do
-    quote do
-      unquote(function_name_helpers())
-      unquote(callback_helpers())
-      import Expression.Callbacks.Standard
-      import Expression.Callbacks.EvalHelpers
+  alias Expression.Callbacks.Standard
+
+  @reserved_words ~w[and if or not]
+
+  @doc """
+  Convert a string function name into an atom meant to handle
+  that function
+
+  Reserved words such as `and`, `if`, and `or` are automatically suffixed
+  with an `_` underscore.
+  """
+  def atom_function_name(function_name) when function_name in @reserved_words,
+    do: atom_function_name("#{function_name}_")
+
+  def atom_function_name(function_name) do
+    String.to_atom(function_name)
+  end
+
+  @doc """
+  Handle a function call while evaluating the AST.
+
+  Handlers in this module are either:
+
+  1. The function name as is
+  2. The function name with an underscore suffix if the function name is a reserved word
+  3. The function name suffixed with `_vargs` if the takes a variable set of arguments
+  """
+  @spec handle(module :: module, function_name :: binary, arguments :: [any], context :: map) ::
+          {:ok, any} | {:error, :not_implemented}
+  def handle(module \\ Standard, function_name, arguments, context) do
+    case implements(module, function_name, arguments) do
+      {:exact, module, function_name, _arity} ->
+        {:ok, apply(module, function_name, [context] ++ arguments)}
+
+      {:vargs, module, function_name, _arity} ->
+        {:ok, apply(module, function_name, [context, arguments])}
+
+      {:error, reason} ->
+        {:error, reason}
     end
   end
 
-  def function_name_helpers do
-    quote do
-      @reserved_words ~w[and if or not]
+  def implements(module \\ Standard, function_name, arguments) do
+    exact_function_name = atom_function_name(function_name)
+    vargs_function_name = atom_function_name("#{function_name}_vargs")
+
+    Code.ensure_loaded!(Standard)
```

Comment:
do we need to document what's going on here?

Reply:
Actually I’m not even sure why this is needed yet

Reply:
oh wait maybe I understand now, otherwise the function exported function doesn't work?

Reply:
I think we can move this to the macro on line 96 to prevent it being called on every function call/evaluation?